## Getting started with NLP and Sentiment Analysis ##

*Notebook based off [Sentiment Analysis for Exploratory Data Analysis](https://programminghistorian.org/en/lessons/sentiment-analysis) by Zöe Wilkinson Saldaña with additional info by [Parul Pandey](https://medium.com/analytics-vidhya/simplifying-social-media-sentiment-analysis-using-vader-in-python-f9e6ec6fc52f).*

Natural Language Processing (NLP) describes a broad range of techniques that involve applying computational analytical methods to textual content. ("Natural language" in this context just means language spoken/written by humans as opposed to a programming language, like Python). In this unit of the course, we'll be exploring many of the most popular NLP techniques, beginning with sentiment analysis. 

Sentiment analysis is a method of quantifing the "sentiment," or emotional intensity, of words and/or phrases within a text. Some sentiment analysis tools, including the one we'll be working with today, also factor in the emotional weight of other features of language, such as punctuation marks or emojis. In general, sentiment analysis tools process a unit of text (a sentence, a paragraph, a book, an email, a song, a Tweet, etc.) and output a set of scores, or other classifications, that indicate whether that unit of text conveys a positive or negative sentiment (according to the particular algorithm employed). Some tools go as far as to quantify the *degree* of positivity or degree of negativity within a text. 

How might this be helpful? A researcher interested in attitudes towards a political event, for example, might use sentiment analysis to characterize how people describe that event on Twitter. Combined with geographic data, sentiment analysis can be used to make comparisons across regions. Combined with demographic data, sentiment analysis can be used to understand how different groups of people view any particular event (or issue, or individual). Sntiment analysis can be easily scaled up, which makes it possible to analyse hundreds of thousands or even millions of speech events.

Like any computational tool, sentiment analysis has a number of limitations that must be taken into account. We'll explore some of those in our discussion today. But when wielding sentiment analysis *both* critically *and* creatively, it can lead to very interesting results.

### NLTK and VADER ###

You will be using Python along with a few tools from [NLTK](https://www.nltk.org/) (short for Natural Language Toolkit) to generate sentiment scores for the corpus that we created in Unit 1: the lyrics of the candidate playlists that we scraped from Genius.com. I've uploaded that corpus to Canvas. 

NLTK is a collection of Python libraries and tools that help researchers apply computational methods to texts. It's been in development since 2001--so, almost as old as you!-- and it's used very widely in the field of NLP. The tools included in NLTK range from methods of breaking up text into smaller pieces, to identifying whether a word belongs in a given language, to sample corpora (that's the plural of corpus) that researchers can use for training and development purposes. We'll be using NLTK a lot in the coming weeks. As with the previous unit, I'll introduce you to some of its features as we require them for our specific goals.   

Today, we will be using one NLTK tool: [VADER](https://github.com/cjhutto/vaderSentiment) (short for Valence Aware Dictionary and sEntiment Reasoner), which generates positive, negative, and neutral sentiment scores for a given textual input.

VADER has a lot of advantages over traditional methods of Sentiment Analysis, including the following:
* It works exceedingly well on social media type text, yet readily generalizes to multiple domains;
* It doesn’t require any training data but is constructed from a generalizable, valence-based, human-curated gold standard sentiment lexicon;
* It is fast enough to be used online with streaming data, and;
* It does not severely suffer from a speed-performance tradeoff.

The source of the above is a very easy to read paper published by the creaters of VADER library, one of whom use to work at Georgia Tech. You can read the paper [here](http://comp.social.gatech.edu/papers/icwsm14.vader.hutto.pdf).

### Sentiment Lexicons, Sentiment Intenstity, and Context-Awareness ###

Traditionally, sentiment analysis worked by comparing the inputted text to a list of lexical features (i.e. words) that were determined by people to be either positive or negative. These are known as *sentiment lexicons.* (It's possible to create lexicons for other types of language as well; we'll talk about this more in Unit 3, when we discuss modeling in more detail). 

More recently, tools have improved upon the positive/negative binary by offering more fine-tuned distinctions between varying degrees of positivity and negativity. This is known as *sentiment intensity*, and VADER does this well. For example, VADER scores “comfort” moderately positively and “euphoria” extremely positively. 

VADER also attempts to capture and score textual features common in informal online text such as capitalizations, exclamation points, and emoticons, as shown in the table below:

![VADER table](https://programminghistorian.org/images/sentiment-analysis/sentiment-analysis1.png)

### Caveat Emptor! ###

Like any text analysis tool, VADER should be evaluated critically and in the context of the assumptions it makes about communication. VADER was developed to analyse English language microblogging and social media sites (especially Twitter). This context is more informal than, for instance, political speeches; and more contemporary than, for instance, Shakespeare. However, VADER was also developed as a general purpose sentiment analyzer, and the authors’ initial study shows it compares favorably against tools that have been trained for specific domains, use specialized lexicons, or resource-heavy machine learning techniques. That said, sentiment analysis continues to struggle to capture complex sentiments like irony, sarcasm, and mockery, when the average reader would be able to make the distinction between the literal text and its intended meaning.

A few more caveats: while VADER is a good general purpose tool for English language texts, VADER only provides partial native support for non-English texts (it detects emojis/capitalization/etc., but not word choice). The developers encourage users to use automatic translation to pre-process non-English texts and then input the results into VADER. There are likely other better tools for non-English langauge texts, although I don't know them off the top of my head. 

### Some examples of hard-to-classify sentences ###

“The premise of the film was great, but it could have been a better movie.”

* What words would you identify as being associated with either positive or negative sentiment?
* Would you say that this sentence has a positive or negative seniment? 

* What are some reasons that this sentence might be tricky for a sentiment analysis tool?

Here's another example:

“The best I can say about the movie is that it was interesting.”

* What words would you identify as being associated with either positive or negative sentiment?
* Would you say that this sentence has a positive or negative seniment? 

* What are some reasons that this sentence might be tricky for a sentiment analysis tool?

### Enough Talk, Time for Action! ###

To use VADER, we need to import the nltk library and download and install the VADER lexicon. You do it like this:

In [1]:
import nltk
nltk.download('vader_lexicon')
# nltk.download('punkt') # took this out of today's lesson 

[nltk_data] Downloading package vader_lexicon to
[nltk_data]     /Users/lauren/nltk_data...
[nltk_data]   Package vader_lexicon is already up-to-date!


True

In order to get a sense of what VADER can do, let’s calculate the sentiment scores for one of the songs we scraped from Genius.com.

The main component of VADER is its SentimentIntensityAnalyzer, so let's import that too:

In [2]:
from nltk.sentiment.vader import SentimentIntensityAnalyzer

/anaconda3/lib/python3.7/site-packages/nltk/twitter/__init__.py:20: UserWarning: The twython library has not been installed. Some functionality from the twitter package will not be available.
  warnings.warn("The twython library has not been installed. "


(You can ignore warning about not having twython installed). 

Technically, SentimentIntensityAnalyzer is a class, which we will use to build our own sentiment analyzer object.

To do so, we call SentimentIntensityAnalyzer() and assign the output - our brand-new sentiment analyzer - to a variable, which we will name ‘sid’.

In [3]:
sid = SentimentIntensityAnalyzer()

By doing this we have given "sid" all of the features of the VADER sentiment analysis object. It has become our sentiment analysis tool, but by a shorter name.

Now, let's open up one of the lyrics files we created:

In [4]:
with open("./lyrics/Aretha-franklin-respect.txt", "r") as file:
    lyrics = file.read()

# and just to be sure, print out what we've loaded in:
lyrics

"What you want, baby, I got it\nWhat you need, do you know I got it?\n\nAll I'm askin' is for a little respect when you come home\n(Just a little bit) Hey baby\n(Just a little bit) when you get home\n(Just a little bit) mister\n(Just a little bit)\n\nI ain't gonna do you wrong while you're gone\nAin't gon' do you wrong 'cause I don't wanna\n\nAll I'm askin' is for a little respect when you come home\n(Just a little bit) Baby\n(Just a little bit) When you get home\n(Just a little bit) Yeah\n(Just a little bit)\n\nI'm about to give you all of my money\nAnd all I'm askin' in return, honey\nIs to give me my propers when you get home\n\n(Just a, just a, just a, just a) Yeah, baby\n(Just a, just a, just a, just a) When you get home\n(Just a little bit) Yeah\n(Just a little bit)\n\nOoh, your kisses, sweeter than honey\nAnd guess what? So is my money\n\nAll I want you to do for me, is give it to me when you get home\n(Re, re, re ,re) Yeah baby\n(Re, re, re ,re) Whip it to me\n(Respect, just a 

What we're most interested in is called the ‘polarity score,’ which gives us a score that is either positive or negative (the two poles). This feature is built into VADER and can be requested on demand.

Calling the polarity_scores method on sid with our lyrics (or, for that matter, any string) outputs a dictionary with negative, neutral, positive, and compound scores for the input text. Let's take a look:

In [5]:
scores = sid.polarity_scores(lyrics)

for key in sorted(scores):
    print('{0}: {1}, '.format(key, scores[key]), end='')

compound: 0.9342, neg: 0.035, neu: 0.875, pos: 0.09, 

Amazing! We just performed our first text analysis! 

But how do we analyze the analysis?

VADER collects and scores negative, neutral, and positive words and features (and accounts for factors like negation along the way). The “neg”, “neu”, and “pos” values describe the fraction of weighted scores that fall into each category. In this case, VADER determined our song lyrics to consist of 3.5% negative words/features, 87.5% neutral words/features, and 9% positive words/features. 

VADER also sums all weighted scores to calculate a “compound” value normalized between -1 and 1; this value attempts to describe the overall affect of the entire chunk of text from strongly negative (-1) to strongly positive (1). In this case, the VADER analysis describes the song as strongly positive (.9342). We can think of this value as estimating the overall impression of an average listener when considering the song as a whole.

This [post](https://medium.com/analytics-vidhya/simplifying-social-media-sentiment-analysis-using-vader-in-python-f9e6ec6fc52f) has a bit more about how VADER calculates its scores.  

Let's [listen to the song](https://www.youtube.com/watch?v=6FOUqQt3Kg0) and see if we agree. 

### A Quick Note on Thresholds ###

It can be helpful to set a minimum threshold for positivity or negativity so that you can classify a text either positive or negative. The official VADER documentation suggests a threshold of -0.5 and 0.5, which this song would meet. 

### Determining Appropriate Scope ###

There isn't much calibration that is required of sentiment analysis. That's why I decided to begin this unit with it. But there is one important aspect of calibration to consider when employing sentiment analysis: the unit of the text being analyzed.

In the case of song lyrics, for example, we might want to analyze the entire song as a single unit, or we might want to analyze each line. 

* What are some research questions for which you might want to look at the entire song as a whole?
* What are some research questions for which you might want to look at each line at a time?

Let's redo our sentiment analysis so that we look at each line of the song individually.



In [6]:
# re-intialize VADER
sid = SentimentIntensityAnalyzer()

# then split our song lyrics into lines broken up by newlines 
lines = lyrics.split('\n') # note handy built-in python string function! 

# We add the additional step of iterating through the list of lines and 
# calculating and printing polarity scores for each one.

for line in lines:
        print(line)
        scores = sid.polarity_scores(line)
        for key in scores:
                print('{0}: {1}, '.format(key, scores[key]), end='')
        print()

What you want, baby, I got it
neg: 0.0, neu: 0.794, pos: 0.206, compound: 0.0772, 
What you need, do you know I got it?
neg: 0.0, neu: 1.0, pos: 0.0, compound: 0.0, 

neg: 0.0, neu: 0.0, pos: 0.0, compound: 0.0, 
All I'm askin' is for a little respect when you come home
neg: 0.0, neu: 0.781, pos: 0.219, compound: 0.4228, 
(Just a little bit) Hey baby
neg: 0.0, neu: 1.0, pos: 0.0, compound: 0.0, 
(Just a little bit) when you get home
neg: 0.0, neu: 1.0, pos: 0.0, compound: 0.0, 
(Just a little bit) mister
neg: 0.0, neu: 1.0, pos: 0.0, compound: 0.0, 
(Just a little bit)
neg: 0.0, neu: 1.0, pos: 0.0, compound: 0.0, 

neg: 0.0, neu: 0.0, pos: 0.0, compound: 0.0, 
I ain't gonna do you wrong while you're gone
neg: 0.307, neu: 0.693, pos: 0.0, compound: -0.4767, 
Ain't gon' do you wrong 'cause I don't wanna
neg: 0.307, neu: 0.693, pos: 0.0, compound: -0.4767, 

neg: 0.0, neu: 0.0, pos: 0.0, compound: 0.0, 
All I'm askin' is for a little respect when you come home
neg: 0.0, neu: 0.781, pos: 0

Here you’ll note a much more detailed picture of the sentiment in the song. 

For ease of interpretation, let's sort the lines by compound sentiment:

In [21]:
from collections import OrderedDict
from operator import itemgetter   

lines_and_compound_scores = {}

for line in lines:
    scores = sid.polarity_scores(line)
    
    lines_and_compound_scores.update( {line: scores['compound']} )

ordered_scores = OrderedDict(sorted(lines_and_compound_scores.items(), key = itemgetter(1), reverse = True))

for key in ordered_scores:
    print("Compound score: " + str(ordered_scores[key]) + ". Line: " + key)

Compound score: 0.5106. Line: Ooh, your kisses, sweeter than honey
Compound score: 0.4939. Line: R-E-S-P-E-C-T, take care, TCB, oh
Compound score: 0.4228. Line: All I'm askin' is for a little respect when you come home
Compound score: 0.4228. Line: A little respect
Compound score: 0.3702. Line: (Just a little bit) A little respect
Compound score: 0.296. Line: (Just a, just a, just a, just a) Yeah, baby
Compound score: 0.296. Line: (Re, re, re ,re) Yeah baby
Compound score: 0.2315. Line: (Just a little bit) Yeah
Compound score: 0.0772. Line: What you want, baby, I got it
Compound score: 0.0772. Line: All I want you to do for me, is give it to me when you get home
Compound score: 0.0. Line: What you need, do you know I got it?
Compound score: 0.0. Line: 
Compound score: 0.0. Line: (Just a little bit) Hey baby
Compound score: 0.0. Line: (Just a little bit) when you get home
Compound score: 0.0. Line: (Just a little bit) mister
Compound score: 0.0. Line: (Just a little bit)
Compound score:

In [11]:
# Class exercise: load in each song and find one with strongest positive sentiment and strongest negative. 